sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeFunctions.scala

```diff
@@ -54,3 +60,313 @@ case class CurrentTimestamp() extends LeafExpression {
     System.currentTimeMillis() * 1000L
   }
 }
+
+/**
+ * Abstract class for create time format expressions.
+ */
+abstract class TimeFormatExpression extends UnaryExpression with ExpectsInputTypes {
+  self: Product =>
+
+  protected val factorToMilli: Int
+
+  protected val cntPerInterval: Int
+
+  override def inputTypes: Seq[AbstractDataType] = Seq(TimestampType)
+
+  override def dataType: DataType = IntegerType
+
+  override protected def nullSafeEval(timestamp: Any): Any = {
+    val time = timestamp.asInstanceOf[Long] / 1000
+    val longTime: Long = time + TimeZone.getDefault.getOffset(time)
+    ((longTime / factorToMilli) % cntPerInterval).toInt
+  }
+
+  override def genCode(ctx: CodeGenContext, ev: GeneratedExpressionCode): String = {
+    val tz = classOf[TimeZone].getName
+    defineCodeGen(ctx, ev, (c) =>
+      s"""(${ctx.javaType(dataType)})
+            ((($c / 1000) + $tz.getDefault().getOffset($c / 1000))
+                / $factorToMilli % $cntPerInterval)"""
+    )
+  }
+}
+
+case class Hour(child: Expression) extends TimeFormatExpression {
+
+  override protected val factorToMilli: Int = 1000 * 3600
+
+  override protected val cntPerInterval: Int = 24
+}
+
+case class Minute(child: Expression) extends TimeFormatExpression {
```

Comment:
I thought that Daylight saving time is just a part of it. The given long value depends on the local time difference, doesn't it. In order to calculate the correct value for `Hour` and `Minute` we need the long time value as of UTC, don't we. For seconds it doesn't matter, but there are time differences like 4:30. So `Minutes` depends on the timezone as well. Maybe I'm wrong and I just didn't got your idea, yet.